spec/models/book_spec.rb

```diff
@@ -0,0 +1,73 @@
+require "rails_helper"
+RSpec.describe Book, type: :model do
+  describe "Associations" do
+    it {should belong_to(:category)}
+    it {should belong_to(:publisher)}
+    it {should have_many(:book_details).dependent(:destroy)}
+    it {should have_many(:book_authors).dependent(:destroy)}
+    it {should have_many(:authors).through(:book_authors)}
+    it {should have_many(:rates).dependent(:destroy)}
+    it {should have_many(:users).through(:rates)}
+    it {should have_one_attached(:image)}
+  end
+
+  describe "Validations" do
+    context "with field name" do
+      subject{FactoryBot.build(:book)}
+      it {should validate_presence_of(:name)}
+      it {should validate_length_of(:name).is_at_most(Settings.max_name_book_length)}
+    end
+
+    context "with field description" do
+      it {should validate_presence_of(:description)}
+      it {should validate_length_of(:description).is_at_most(Settings.max_description_length)}
+    end
+  end
+
+  describe "Scope" do
+    let!(:book_1){FactoryBot.create :book, name:"sach_1"}
```

Comment:
phân biệt let và let!

Reply:
ở đây dùng let thôi đc ko

Reply:
- let sẽ có giá trị trong từng khối it và chỉ được gọi lần đầu khi khối it được thực thi sau đó sẽ được cached lại những lần gọi sau trong cùng khối it giá trị sẽ là giá trị được cached
- let! được thực thi trong khối before ngầm do đó nó được gọi trước toàn bộ các khối it và có giá trị giống nhau cho các khối it